core/src/main/scala/kafka/common/InterBrokerSendThread.scala

```diff
@@ -0,0 +1,72 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package kafka.common
+
+import kafka.utils.{ShutdownableThread, ZkUtils}
+import org.apache.kafka.clients.{ClientResponse, NetworkClient, RequestCompletionHandler}
+import org.apache.kafka.common.Node
+import org.apache.kafka.common.requests.AbstractRequest
+import org.apache.kafka.common.utils.Time
+
+
+object InterBrokerSendThread {
+
+}
+
+/**
+ *  Class for inter-broker send thread that utilize a non-blocking network client.
+ */
+class InterBrokerSendThread(name: String,
+                            networkClient: NetworkClient,
+                            requestGenerator: () => Iterable[RequestAndCompletionHandler],
+                            time: Time)
+  extends ShutdownableThread(name, isInterruptible = false) {
+
+  override def doWork() {
+    val now = time.milliseconds()
+    var pollTimeout = Long.MaxValue
+
+    val requestsToSend: Iterable[RequestAndCompletionHandler] = requestGenerator()
+
+    for (request: RequestAndCompletionHandler <- requestsToSend) {
+      val destination = Integer.toString(request.destination.id())
+      val completionHandler = request.handler
+      val clientRequest = networkClient.newClientRequest(destination,
+        request.request,
+        now,
+        true,
+        completionHandler)
+
+      if (networkClient.ready(request.destination, now)) {
+        networkClient.send(clientRequest, now)
+      } else {
+        val disConnectedResponse: ClientResponse = new ClientResponse(clientRequest.makeHeader(request.request.desiredOrLatestVersion()),
```

Comment:
I agree with @junrao that this is inefficient. The producer model where we have a background sender thread and one queue per broker is more efficient, since it avoids busy waiting. However, we agreed that we can make this improvement in the future, after the initial integration is done.

I filed this ticket to keep track of the work item: https://confluentinc.atlassian.net/browse/CPKAFKA-673